lib/pbench/agent/tool_data_sink.py

```diff
@@ -1295,19 +1300,45 @@ def execute(self):
                 )
             self.logger.debug("published %s", self._to_client_channel)
 
-            for data in self._from_client_chan.fetch_json(self.logger):
-                ret_val = self._valid_data(data)
-                if ret_val is None:
+            # Begin listening for client state signals, and react/respond
+            # accordingly upon signal receipt
+            for signal in self.sig_resp.listen():
+                client = signal.publisher_id
+                action = signal.event
+                self.logger.debug(
+                    "client %r, action %r, metadata %r", client, action, signal.metadata
+                )
+                if action == "shutdown":
+                    # This simply means that a SignalExporter/Client we
+                    # were listening to has shut down, no action needed
+                    self.logger.debug("Client %s has shut down", client)
                     continue
-                action, directory, args = ret_val[0], ret_val[1], ret_val[2]
-                self.execute_action(action, directory, args, data)
-            self.logger.info("Tool Data Sink terminating")
+                if action == "initialization":
+                    self.sig_resp.srespond(signal)
+                    continue
+                if not signal.metadata:
+                    self.logger.debug(
```

Comment:
Should this really be a `debug`? It means a remote is mismatched or broken, right? Unlikely as it (hopefully) is, shouldn't this be at least a `warning`?

Or better, this should probably just be removed, as it's redundant with the more complete check in `_is_valid_data` immediately below ...

Reply:
I think it's good to keep the two separate, as `_is_valid_data` simply checks the validity of a passed-in metadata (assumes it exists), so a check to make sure there is ANY data would still be required. I think you are right about changing it to a `warning`, though, as that keeps the logging in-line with how `_is_valid_data` reports malformed data.